### ISSUE TITLE
flows long time delay when use network interface live captrue

### ISSUE THREAD

**Issue by missyoyo** [nfstream/nfstream#106]
**Describe the bug**
Hi, I have some performance test recently and find a issue, I use Apache JMeter to get some http test(http object 100KB no keep alive).I send 5863 http get in 10 seconds and use nfstream live network interface captrue,I get 5869 results use nfstream, seem total count has no problem,but unfortunately I start test at 21:00 but last record received at 22:01 take 1 hour. I have some other softe real time monitor tcp and http connections and also use tcpdump captrue test pcap too.
1. tcp connections get form wireshark(5863 same as JMeter)
![图片](https://user-images.githubusercontent.com/19371588/155869321-449ac340-1905-413c-8fbf-83903a9a6393.png)
2. Resulte get from nfstream(5869 with long time delay)
![图片](https://user-images.githubusercontent.com/19371588/155869399-fa356dc0-8b80-454a-8906-274d5f211bee.png)
3.tcp monitor from other soft(6068 with 5 minutes delay)
![图片](https://user-images.githubusercontent.com/19371588/155869414-c4cbe927-3f33-4224-a173-c75ce3538a3e.png)

**Desktop (please complete the following information):**
 - OS: [Ubuntu 18.04]
 - Python version [3.6.9]
 - NFStream version [6.4.2]

**Additional context**
When I use pacp as nfstream inpute seems no problem.
message example:
{"id": 4388, "expiration_id": 0, "src_ip": "142.250.73.212", "src_mac": "00:0c:29:7b:51:38", "src_oui": "00:0c:29", "src_port": 51664, "dst_ip": "142.250.73.2", "dst_mac": "00:0c:29:c1:d6:c6", "dst_oui": "00:0c:29", "dst_port": 80, "protocol": 6, "ip_version": 4, "vlan_id": 0, "tunnel_id": 0, "bidirectional_first_seen_ms": 1645880543390, "bidirectional_last_seen_ms": 1645880550350, "bidirectional_duration_ms": 6960, "bidirectional_packets": 146, "bidirectional_bytes": 143166, "src2dst_first_seen_ms": 1645880543390, "src2dst_last_seen_ms": 1645880550349, "src2dst_duration_ms": 6959, "src2dst_packets": 50, "src2dst_bytes": 3379, "dst2src_first_seen_ms": 1645880543390, "dst2src_last_seen_ms": 1645880550350, "dst2src_duration_ms": 6960, "dst2src_packets": 96, "dst2src_bytes": 139787}
I can see first see packet at 2022-02-26 21:02:23 and last packet see 2022-02-26 21:02:30 and only 6.9 seconds, But I am sure there always some resulte take very long delay. in this time(1514 resultes take 1 hour delay,744 resultes take 30 minute delay).
I see http client send FIN/ACK to http server. Is that mean client end this tcp connections normally and nfstream will get resulte immediately or need wait tcp time out?
I also set  active_timeout very short.
![图片](https://user-images.githubusercontent.com/19371588/155869798-fa48c4ae-03e5-4238-a1d3-d172333eaba3.png)

![图片](https://user-images.githubusercontent.com/19371588/155869798-fa48c4ae-03e5-4238-a1d3-d172333eaba3.png)

**Comment by drnpkr** [comment]
Hi @missyoyo,

I think I know what could be the problem. Can you please try using different idle and active timeouts and get back with the results/observations?

Thanks

**Comment by aouinizied** [comment]
@missyoyo I will try to explain what is happening there:
* nfstream starts normally and spawn several meters on your host.
* each meter will process a subpart of your traffic based on kernel flow hash based load balancing mechanism (as you are doing live capture on Linux system).

In practice, this means that each packet belonging to a bidirectional flow will be treated by the same meter. let's take back some flows from your example and to simplify things let's suppose that we have only two flows and two meters:

* **bi_flow_1**: 142.250.73.212:51044:142.250.73.2:80:6
* **bi_flow_2**: 142.250.73.212:51045:142.250.73.2:80:6

All flow 1 packets are processed by meter 1 and all flow 2 packets are processed by meter 2.
When the last flow_2 packet hits meter 2, we will see the record when meter 2 expires it based on idle timeout.
So if last packets time is 12:00:00 and idle timeout is set to 10 secs, the flow expires only when internal meter clock is >= then 12:00:10. Internal meter clock is within current version updated only with packets. Consequently, in the previous illustration you will see the record at 12:00:10 only if there is packets processed by this meter within this time range. If the only packet coming to this meter after the last flow 2 packet is at 13:00:00, you will see the record at 13:00:00.

Note that this is the behavior on Linux platform for live capture. On macOS , Windows or on offline, behavior is slightly different.

The main idea behind such a logic is that there is always a packet flowing somewhere each second.
For your specific case I guess no packets are sent after each test flow.

This behavior will be changed in next release and this will obviously fix what you are observing.

Zied

**Comment by zhangineer2** [comment]
was this enhancement ever added ? I didn't quite understand the explanation. in the example given:
@aouinizied you mentioned that `When the last flow_2 packets hits meter 2, we will see the record when meter 2 expires it based on idle timeout`, but what about flow 1 ?

in a test that I did, where I had 20 flows in total. when i sent all 20 flows in a 30 seconds window, I had to wait for 3 minutes to see any records, and when the records came in, there were 10 of them bundled together all sent at once. Is this because all 10 flows are tracked with a single meter ?

but why am I only seeing this after 3 minutes ?
